### PR TITLE
QML locator filter framework for QField plugin

### PR DESCRIPTION
**Pull request by nirvn**
_Much wisdom gained from these lines of code._

The PR adds a framework that allows for plugin authors to integrate into the search bar through a QML-driven locator filter.

Here's a sample plugin registering a QML locator filter being declared:

```qml
Item {
  id: plugin

  Component.onCompleted: {
    locatorBridge.registerQFieldLocatorFilter(testLocatorFilter);
  }

  QFieldLocatorFilter {
    id: testLocatorFilter

    name: "testlocator"
    displayName: "test locator!!!"
    prefix: "123"

    source: Qt.resolvedUrl('testlocator.qml')
    
    function triggerResult(result) {
      console.log(result)
    }
  }
}
```

The QFieldLocatorFilter item *lives in the same thread and context* as QField's mainWindow, and therefore can access the map canvas et al within the `triggerResult` function.

Notice the source property of QFieldLocatorFilter. This is a sidecar QML component file that lives next to the main plugin QML (i.e. main.qml for app-wide plugins and my_project_file_name.qml for project plugins). That component must be declared in its own file as everything in there will be happening in a separate QML engine altogether, in a separate thread.

~~Also notice the use of a new property named `pluginFolder`. This was added so plugins can load extra resource from their folder. To make use of it, simply declare a property var pluginFolder at the root of your plugin root Item {}.~~ Note the use Qt.resolveUrl('my_file.qml') to resolve relative paths against the source file of the QML being loaded for the source property.

Moving on, here's a sample example of the source file (named testlocator.qml in the example above):

```qml
Item {
  signal prepareResult(var details)
  signal fetchResultsEnded()
  
  Timer {
    id: timer
    interval: 2000
    running: false
    repeat: false
    
    onTriggered: {
      let details = {
        "userData": GeometryUtils.point(0, 0),
        "displayString": "My title",
        "description": "My description...",
        "score": 0.5,
        "group": "My group",
        "groupScore": 0.5
      };
      prepareResult(details);
      fetchResultsEnded();
    }
  }
  
  function fetchResults(string, context) {
    timer.start();
  }
}
```

The use of the Timer here is merely to demonstrate how asynchronous result gathering happens. First, the function fetchResults(string, context) will be called. *If the results can be computed right away*, simply emit prepareResult(details) as many time as you want, and end the function by emitting fetchResultEnded(). However, it the very likely scenario the result fetching is waiting for some data (e.g. waiting for the response of an xmlhttprequest), the locator filter itself will wait until the fetchResultEnded() is emitted.

When all is done and understood, you'll get a beautiful integration:

[Screencast](https://github.com/user-attachments/assets/63ae86c2-b59c-4a7f-ae73-1c39ee2bafe8)

Here's a sample plugin that adds a GeoMapFish service into the search bar:

[qfield-geomapfish-locator-v1.1.zip](https://github.com/user-attachments/files/17893072/qfield-geomapfish-locator-v1.1.zip)

Another sample plugin that adds OSM nominatim service into the search bar:

[qfield-nominatim-locator-v1.1.zip](https://github.com/user-attachments/files/17910619/qfield-nominatim-locator-v1.1.zip)

The framework also allows for plugins to add secondary actions to results. For example here, the search result has a secondary option to use the result geometry as destination:

[0001-0220.webm](https://github.com/user-attachments/assets/020ee185-1e8c-4d9a-937b-da025e35a41a)

### CODE REVIEW

File: src/core/locator/qfieldlocatorfilter.cpp [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    return;
+
+  QQmlEngine engine;
+  QQmlComponent component( &engine );
+  component.loadUrl( mSource );
+  QObject *object = component.create();
+  if ( object )
+  {
+    mFetchResultsEnded = false;
+    connect( object, SIGNAL( prepareResult( QVariant ) ), this, SLOT( prepareResult( QVariant ) ) );
```
Comment (m-kuhn) [inline src/core/locator/qfieldlocatorfilter.cpp:124]: Is there a reason we use old-style signals/slots here?

Comment (m-kuhn) [reply src/core/locator/qfieldlocatorfilter.cpp:124]: can we do a qobject_cast?

File: src/core/locator/qfieldlocatorfilter.cpp [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    connect( object, SIGNAL( fetchResultsEnded() ), this, SLOT( fetchResultsEnded() ) );
+
+    QEventLoop loop;
+    connect( object, SIGNAL( fetchResultsEnded() ), &loop, SLOT( quit() ) );
+    connect( feedback, &QgsFeedback::canceled, &loop, &QEventLoop::quit );
+
+    QMetaObject::invokeMethod( object, QStringLiteral( "fetchResults" ).toStdString().c_str(), QVariant( string ), QVariant::fromValue<QgsLocatorContext>( context ) );
+    if ( !mFetchResultsEnded )
+    {
+      loop.exec();
```
Comment (m-kuhn) [inline src/core/locator/qfieldlocatorfilter.cpp:134]: is the interface still responsive here?
i.e. what happens if there's a long-running web request (imagine timeout 60 seconds or  more)?

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 67e53a535025260f5ac7bbebbf90a51a13574058: [**arm64-android**](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5841-a595d105ba51980a4c5fd73595868c3f62a25b57-android-arm64.apk)
<details>
<summary>Other architectures</summary>

- [arm-neon-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5841-a595d105ba51980a4c5fd73595868c3f62a25b57-android-arm-neon.apk)
- [x64-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5841-a595d105ba51980a4c5fd73595868c3f62a25b57-android-x64.apk)
- [x86-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5841-a595d105ba51980a4c5fd73595868c3f62a25b57-android-x86.apk)
- [**all access arm64-android**](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5841-a595d105ba51980a4c5fd73595868c3f62a25b57-android-arm64.apk)
- [all access arm-neon-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5841-a595d105ba51980a4c5fd73595868c3f62a25b57-android-arm-neon.apk)
- [all access x64-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5841-a595d105ba51980a4c5fd73595868c3f62a25b57-android-x64.apk)
- [all access x86-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5841-a595d105ba51980a4c5fd73595868c3f62a25b57-android-x86.apk)
</details>

File: src/core/locator/qfieldlocatorfilter.cpp (hunk 2 above)
Comment (nirvn) [reply src/core/locator/qfieldlocatorfilter.cpp:134]: Yep, the interface isn't impacted with what's happening here as this is running in a cloned filter _off_ the main QML engine / UI thread.

File: src/core/qgsquick/qgsquickmapsettings.h [hunk 3]
```diff
@@ -154,7 +154,7 @@ class QFIELD_CORE_EXPORT QgsQuickMapSettings : public QObject
     QgsRectangle extent() const;
 
     //! \copydoc QgsMapSettings::setExtent()
-    void setExtent( const QgsRectangle &extent, bool handleMargins = false );
+    Q_INVOKABLE void setExtent( const QgsRectangle &extent, bool handleMargins = false );
```
Comment (m-kuhn) [inline src/core/qgsquick/qgsquickmapsettings.h:157]: Just wondering: property instead?

Comment (nirvn) [reply src/core/qgsquick/qgsquickmapsettings.h:157]: @m-kuhn , the property is there, but it ignores the 2nd parameter (i.e. handleMargins), whereas I want our locators to not re-center onto a covered part of the canvas :)

**Pull request merged**